allauth/account/forms.py

```diff
@@ -335,13 +336,23 @@ def __init__(self, *args, **kwargs):
         )
 
     def clean_username(self):
-        value = self.cleaned_data["username"]
-        value = get_adapter().clean_username(value)
+        if self.cleaned_data["username"]:
+            value = self.cleaned_data["username"]
+        else:
+            value = get_adapter().clean_username(value)
+        users = User.objects.filter(username__iexact=value)
+        if users:
+            raise forms.ValidationError("A link to activate your account has been emailed to the address provided")
```

Comment:
Currently, the user would see an error "username taken". So, that indeed leaks information. Now, you would see a different error message if the username is taken -- how does this leak less information? One can come to the same conclusion that a username is taken. 

Also, the error message is simply not stating what truly happens. Suppose I signed up with a username before, and forgot about it. Then, 1 month later, I signup again using the same username. I am presented that error message, and will spend looking at my mailbox for that mail to come in, but no mail will arrive as this is only changing the message itself?

Reply:
Good point an error message about an email link being sent does not make sense for a username. Especially if the email verification setting is None. Or email address is not used for registration at all. 

The point of these error messages is to make it less obvious if the username/email was taken or not. In the case of emails not being leaked it does not explicitly say that the email is taken and makes someone think even if the email address does or does not exist, an email will "appear" to be sent for activation.

Maybe we could just change the duplicate error message for usernames to be "username is invalid" even if there is a duplicate username already being used. It will force the user to use a different username and will not explicitly tell if the username typed in already exists to help prevent brute force attacks. 

Your scenario makes sense. I feel as though that would be a rare occurrence though. This whole thing falls into harvesting and enumeration categories of security. We really need a middle-ground between good error messaging and security without there hopefully being a complete trade off for one or the other. 

It would be cool to be able to allow users to customize these forms.py validation errors without them being hard coded. But not sure how to do that myself.  

You could also change the wording for emails perhaps to be "A link to activate your account may or may not be emailed to the address provided" so that it satisfies your scenario for emails perhaps?